# DCS 1100
# Deep Learning for Automatically Tagging Medals

# Before running the code

Copy the Medals Project folder from Microwave to your personal computer.

Windows only: Install git
https://git-scm.com/download/win

Both Mac and Windows: In a command terminal, "cd" to the project folder and execute these commands

    pip install tensorflow
    git clone https://github.com/tensorflow/tensorflow.git


## Training
The following code trains the last layer of an ANN using tagged images.
Tagging is implicitly done by putting medal images in appropriate folders.<br>
<strong>Caution:</strong> Training is computationally intensive. Your browser will freeze. <u>Do not kill the browser.</u> Use another browser if you need to use the Internet. (Ignore logging errors.)

In [1]:
import tensorflow
%run C://Users/Alessandra/Anaconda3/MedalsProject/tensorflow/tensorflow/examples/image_retraining/retrain.py --image_dir training_set

Not extracting or downloading files, model already present in disk
Model path:  /tmp/imagenet\classify_image_graph_def.pb
INFO:tensorflow:Looking for images in 'animal'
INFO:tensorflow:Looking for images in 'blank'
INFO:tensorflow:Looking for images in 'building'
INFO:tensorflow:Looking for images in 'child'
INFO:tensorflow:Looking for images in 'figure'
INFO:tensorflow:Looking for images in 'group'
INFO:tensorflow:Looking for images in 'mythical'
INFO:tensorflow:Looking for images in 'object'
INFO:tensorflow:Looking for images in 'plants'
INFO:tensorflow:Looking for images in 'portrait'
INFO:tensorflow:Looking for images in 'symbol'
INFO:tensorflow:Looking for images in 'text'
INFO:tensorflow:100 bottleneck files created.
INFO:tensorflow:200 bottleneck files created.
INFO:tensorflow:300 bottleneck files created.
INFO:tensorflow:400 bottleneck files created.
INFO:tensorflow:500 bottleneck files created.
INFO:tensorflow:600 bottleneck files created.
INFO:tensorflow:700 bottleneck files 

INFO:tensorflow:2017-12-08 22:42:35.772370: Step 260: Cross entropy = 1.825106
INFO:tensorflow:2017-12-08 22:42:36.358287: Step 260: Validation accuracy = 27.0% (N=100)
INFO:tensorflow:2017-12-08 22:42:40.986580: Step 270: Train accuracy = 54.0%
INFO:tensorflow:2017-12-08 22:42:40.988082: Step 270: Cross entropy = 1.697541
INFO:tensorflow:2017-12-08 22:42:41.442405: Step 270: Validation accuracy = 25.0% (N=100)
INFO:tensorflow:2017-12-08 22:42:45.897075: Step 280: Train accuracy = 51.0%
INFO:tensorflow:2017-12-08 22:42:45.898577: Step 280: Cross entropy = 1.712433
INFO:tensorflow:2017-12-08 22:42:46.342892: Step 280: Validation accuracy = 22.0% (N=100)
INFO:tensorflow:2017-12-08 22:42:50.818077: Step 290: Train accuracy = 48.0%
INFO:tensorflow:2017-12-08 22:42:50.819579: Step 290: Cross entropy = 1.683533
INFO:tensorflow:2017-12-08 22:42:51.241879: Step 290: Validation accuracy = 28.0% (N=100)
INFO:tensorflow:2017-12-08 22:42:55.988303: Step 300: Train accuracy = 48.0%
INFO:tensorflow:

INFO:tensorflow:2017-12-08 22:45:27.127262: Step 590: Validation accuracy = 21.0% (N=100)
INFO:tensorflow:2017-12-08 22:45:31.104593: Step 600: Train accuracy = 53.0%
INFO:tensorflow:2017-12-08 22:45:31.106094: Step 600: Cross entropy = 1.488247
INFO:tensorflow:2017-12-08 22:45:31.558916: Step 600: Validation accuracy = 28.0% (N=100)
INFO:tensorflow:2017-12-08 22:45:36.233754: Step 610: Train accuracy = 56.0%
INFO:tensorflow:2017-12-08 22:45:36.235256: Step 610: Cross entropy = 1.424979
INFO:tensorflow:2017-12-08 22:45:36.647549: Step 610: Validation accuracy = 25.0% (N=100)
INFO:tensorflow:2017-12-08 22:45:41.236321: Step 620: Train accuracy = 65.0%
INFO:tensorflow:2017-12-08 22:45:41.237823: Step 620: Cross entropy = 1.284375
INFO:tensorflow:2017-12-08 22:45:41.660123: Step 620: Validation accuracy = 26.0% (N=100)
INFO:tensorflow:2017-12-08 22:45:46.318939: Step 630: Train accuracy = 61.0%
INFO:tensorflow:2017-12-08 22:45:46.320941: Step 630: Cross entropy = 1.357951
INFO:tensorflow:

INFO:tensorflow:2017-12-08 22:48:08.613414: Step 930: Train accuracy = 61.0%
INFO:tensorflow:2017-12-08 22:48:08.615416: Step 930: Cross entropy = 1.335644
INFO:tensorflow:2017-12-08 22:48:09.055728: Step 930: Validation accuracy = 34.0% (N=100)
INFO:tensorflow:2017-12-08 22:48:13.941706: Step 940: Train accuracy = 67.0%
INFO:tensorflow:2017-12-08 22:48:13.943708: Step 940: Cross entropy = 1.263296
INFO:tensorflow:2017-12-08 22:48:14.465578: Step 940: Validation accuracy = 26.0% (N=100)
INFO:tensorflow:2017-12-08 22:48:19.666280: Step 950: Train accuracy = 68.0%
INFO:tensorflow:2017-12-08 22:48:19.667781: Step 950: Cross entropy = 1.301868
INFO:tensorflow:2017-12-08 22:48:20.402804: Step 950: Validation accuracy = 37.0% (N=100)
INFO:tensorflow:2017-12-08 22:48:26.206935: Step 960: Train accuracy = 56.0%
INFO:tensorflow:2017-12-08 22:48:26.208436: Step 960: Cross entropy = 1.483755
INFO:tensorflow:2017-12-08 22:48:26.641744: Step 960: Validation accuracy = 25.0% (N=100)
INFO:tensorflow:

INFO:tensorflow:2017-12-08 22:51:13.217188: Step 1260: Train accuracy = 76.0%
INFO:tensorflow:2017-12-08 22:51:13.218690: Step 1260: Cross entropy = 1.164087
INFO:tensorflow:2017-12-08 22:51:13.655501: Step 1260: Validation accuracy = 28.0% (N=100)
INFO:tensorflow:2017-12-08 22:51:18.018149: Step 1270: Train accuracy = 63.0%
INFO:tensorflow:2017-12-08 22:51:18.019651: Step 1270: Cross entropy = 1.261551
INFO:tensorflow:2017-12-08 22:51:18.472973: Step 1270: Validation accuracy = 16.0% (N=100)
INFO:tensorflow:2017-12-08 22:51:23.514064: Step 1280: Train accuracy = 71.0%
INFO:tensorflow:2017-12-08 22:51:23.516563: Step 1280: Cross entropy = 1.267327
INFO:tensorflow:2017-12-08 22:51:24.016418: Step 1280: Validation accuracy = 23.0% (N=100)
INFO:tensorflow:2017-12-08 22:51:29.447798: Step 1290: Train accuracy = 60.0%
INFO:tensorflow:2017-12-08 22:51:29.449299: Step 1290: Cross entropy = 1.256585
INFO:tensorflow:2017-12-08 22:51:29.926638: Step 1290: Validation accuracy = 28.0% (N=100)
INFO

INFO:tensorflow:2017-12-08 22:54:11.507191: Step 1590: Train accuracy = 73.0%
INFO:tensorflow:2017-12-08 22:54:11.509194: Step 1590: Cross entropy = 0.981693
INFO:tensorflow:2017-12-08 22:54:11.947004: Step 1590: Validation accuracy = 25.0% (N=100)
INFO:tensorflow:2017-12-08 22:54:16.231553: Step 1600: Train accuracy = 70.0%
INFO:tensorflow:2017-12-08 22:54:16.240560: Step 1600: Cross entropy = 1.022552
INFO:tensorflow:2017-12-08 22:54:16.681874: Step 1600: Validation accuracy = 26.0% (N=100)
INFO:tensorflow:2017-12-08 22:54:21.075501: Step 1610: Train accuracy = 62.0%
INFO:tensorflow:2017-12-08 22:54:21.077002: Step 1610: Cross entropy = 1.268830
INFO:tensorflow:2017-12-08 22:54:21.528823: Step 1610: Validation accuracy = 23.0% (N=100)
INFO:tensorflow:2017-12-08 22:54:26.442321: Step 1620: Train accuracy = 67.0%
INFO:tensorflow:2017-12-08 22:54:26.445328: Step 1620: Cross entropy = 1.237318
INFO:tensorflow:2017-12-08 22:54:26.948180: Step 1620: Validation accuracy = 27.0% (N=100)
INFO

INFO:tensorflow:2017-12-08 22:57:04.064242: Step 1920: Train accuracy = 71.0%
INFO:tensorflow:2017-12-08 22:57:04.065743: Step 1920: Cross entropy = 1.069986
INFO:tensorflow:2017-12-08 22:57:04.512061: Step 1920: Validation accuracy = 31.0% (N=100)
INFO:tensorflow:2017-12-08 22:57:08.938211: Step 1930: Train accuracy = 72.0%
INFO:tensorflow:2017-12-08 22:57:08.939712: Step 1930: Cross entropy = 1.019593
INFO:tensorflow:2017-12-08 22:57:09.377524: Step 1930: Validation accuracy = 24.0% (N=100)
INFO:tensorflow:2017-12-08 22:57:13.772651: Step 1940: Train accuracy = 68.0%
INFO:tensorflow:2017-12-08 22:57:13.774653: Step 1940: Cross entropy = 1.083336
INFO:tensorflow:2017-12-08 22:57:14.231478: Step 1940: Validation accuracy = 25.0% (N=100)
INFO:tensorflow:2017-12-08 22:57:19.478212: Step 1950: Train accuracy = 67.0%
INFO:tensorflow:2017-12-08 22:57:19.480214: Step 1950: Cross entropy = 1.007302
INFO:tensorflow:2017-12-08 22:57:20.174710: Step 1950: Validation accuracy = 24.0% (N=100)
INFO

INFO:tensorflow:2017-12-08 23:00:00.057545: Step 2250: Train accuracy = 82.0%
INFO:tensorflow:2017-12-08 23:00:00.060046: Step 2250: Cross entropy = 0.888955
INFO:tensorflow:2017-12-08 23:00:00.569409: Step 2250: Validation accuracy = 27.0% (N=100)
INFO:tensorflow:2017-12-08 23:00:05.386838: Step 2260: Train accuracy = 75.0%
INFO:tensorflow:2017-12-08 23:00:05.389340: Step 2260: Cross entropy = 0.942580
INFO:tensorflow:2017-12-08 23:00:05.821146: Step 2260: Validation accuracy = 17.0% (N=100)
INFO:tensorflow:2017-12-08 23:00:10.527496: Step 2270: Train accuracy = 77.0%
INFO:tensorflow:2017-12-08 23:00:10.529497: Step 2270: Cross entropy = 0.955720
INFO:tensorflow:2017-12-08 23:00:10.999845: Step 2270: Validation accuracy = 23.0% (N=100)
INFO:tensorflow:2017-12-08 23:00:15.637145: Step 2280: Train accuracy = 67.0%
INFO:tensorflow:2017-12-08 23:00:15.638647: Step 2280: Cross entropy = 1.032230
INFO:tensorflow:2017-12-08 23:00:16.069953: Step 2280: Validation accuracy = 23.0% (N=100)
INFO

INFO:tensorflow:2017-12-08 23:03:13.563480: Step 2580: Train accuracy = 76.0%
INFO:tensorflow:2017-12-08 23:03:13.565481: Step 2580: Cross entropy = 0.944464
INFO:tensorflow:2017-12-08 23:03:14.262977: Step 2580: Validation accuracy = 30.0% (N=100)
INFO:tensorflow:2017-12-08 23:03:22.531361: Step 2590: Train accuracy = 74.0%
INFO:tensorflow:2017-12-08 23:03:22.532864: Step 2590: Cross entropy = 0.988598
INFO:tensorflow:2017-12-08 23:03:23.364455: Step 2590: Validation accuracy = 19.0% (N=100)
INFO:tensorflow:2017-12-08 23:03:31.469223: Step 2600: Train accuracy = 79.0%
INFO:tensorflow:2017-12-08 23:03:31.471225: Step 2600: Cross entropy = 0.888942
INFO:tensorflow:2017-12-08 23:03:32.291809: Step 2600: Validation accuracy = 28.0% (N=100)
INFO:tensorflow:2017-12-08 23:03:40.196774: Step 2610: Train accuracy = 81.0%
INFO:tensorflow:2017-12-08 23:03:40.198776: Step 2610: Cross entropy = 0.806941
INFO:tensorflow:2017-12-08 23:03:40.898273: Step 2610: Validation accuracy = 24.0% (N=100)
INFO

INFO:tensorflow:2017-12-08 23:06:10.273700: Step 2910: Train accuracy = 79.0%
INFO:tensorflow:2017-12-08 23:06:10.275201: Step 2910: Cross entropy = 0.874838
INFO:tensorflow:2017-12-08 23:06:10.679489: Step 2910: Validation accuracy = 27.0% (N=100)
INFO:tensorflow:2017-12-08 23:06:14.993559: Step 2920: Train accuracy = 73.0%
INFO:tensorflow:2017-12-08 23:06:14.995060: Step 2920: Cross entropy = 0.981840
INFO:tensorflow:2017-12-08 23:06:15.517432: Step 2920: Validation accuracy = 16.0% (N=100)
INFO:tensorflow:2017-12-08 23:06:19.613347: Step 2930: Train accuracy = 73.0%
INFO:tensorflow:2017-12-08 23:06:19.615349: Step 2930: Cross entropy = 0.967475
INFO:tensorflow:2017-12-08 23:06:20.015633: Step 2930: Validation accuracy = 30.0% (N=100)
INFO:tensorflow:2017-12-08 23:06:24.002971: Step 2940: Train accuracy = 73.0%
INFO:tensorflow:2017-12-08 23:06:24.004473: Step 2940: Cross entropy = 0.964402
INFO:tensorflow:2017-12-08 23:06:24.413263: Step 2940: Validation accuracy = 25.0% (N=100)
INFO

INFO:tensorflow:2017-12-08 23:09:03.528405: Step 3240: Train accuracy = 74.0%
INFO:tensorflow:2017-12-08 23:09:03.529906: Step 3240: Cross entropy = 0.925925
INFO:tensorflow:2017-12-08 23:09:03.980226: Step 3240: Validation accuracy = 24.0% (N=100)
INFO:tensorflow:2017-12-08 23:09:09.018312: Step 3250: Train accuracy = 73.0%
INFO:tensorflow:2017-12-08 23:09:09.021317: Step 3250: Cross entropy = 0.884475
INFO:tensorflow:2017-12-08 23:09:09.532177: Step 3250: Validation accuracy = 29.0% (N=100)
INFO:tensorflow:2017-12-08 23:09:14.690349: Step 3260: Train accuracy = 69.0%
INFO:tensorflow:2017-12-08 23:09:14.692350: Step 3260: Cross entropy = 0.970429
INFO:tensorflow:2017-12-08 23:09:15.237739: Step 3260: Validation accuracy = 33.0% (N=100)
INFO:tensorflow:2017-12-08 23:09:19.997125: Step 3270: Train accuracy = 79.0%
INFO:tensorflow:2017-12-08 23:09:19.999128: Step 3270: Cross entropy = 0.849048
INFO:tensorflow:2017-12-08 23:09:20.480469: Step 3270: Validation accuracy = 24.0% (N=100)
INFO

INFO:tensorflow:2017-12-08 23:12:22.932890: Step 3570: Train accuracy = 81.0%
INFO:tensorflow:2017-12-08 23:12:22.934392: Step 3570: Cross entropy = 0.806447
INFO:tensorflow:2017-12-08 23:12:23.491288: Step 3570: Validation accuracy = 28.0% (N=100)
INFO:tensorflow:2017-12-08 23:12:28.884126: Step 3580: Train accuracy = 76.0%
INFO:tensorflow:2017-12-08 23:12:28.885628: Step 3580: Cross entropy = 0.871221
INFO:tensorflow:2017-12-08 23:12:29.490557: Step 3580: Validation accuracy = 25.0% (N=100)
INFO:tensorflow:2017-12-08 23:12:34.477107: Step 3590: Train accuracy = 70.0%
INFO:tensorflow:2017-12-08 23:12:34.478608: Step 3590: Cross entropy = 0.901659
INFO:tensorflow:2017-12-08 23:12:34.895404: Step 3590: Validation accuracy = 27.0% (N=100)
INFO:tensorflow:2017-12-08 23:12:39.993547: Step 3600: Train accuracy = 70.0%
INFO:tensorflow:2017-12-08 23:12:39.995048: Step 3600: Cross entropy = 0.943172
INFO:tensorflow:2017-12-08 23:12:40.486898: Step 3600: Validation accuracy = 24.0% (N=100)
INFO

INFO:tensorflow:2017-12-08 23:15:39.030219: Step 3900: Train accuracy = 71.0%
INFO:tensorflow:2017-12-08 23:15:39.031721: Step 3900: Cross entropy = 0.916320
INFO:tensorflow:2017-12-08 23:15:39.522069: Step 3900: Validation accuracy = 29.0% (N=100)
INFO:tensorflow:2017-12-08 23:15:45.064013: Step 3910: Train accuracy = 66.0%
INFO:tensorflow:2017-12-08 23:15:45.065515: Step 3910: Cross entropy = 1.039927
INFO:tensorflow:2017-12-08 23:15:45.500824: Step 3910: Validation accuracy = 22.0% (N=100)
INFO:tensorflow:2017-12-08 23:15:50.047560: Step 3920: Train accuracy = 79.0%
INFO:tensorflow:2017-12-08 23:15:50.049061: Step 3920: Cross entropy = 0.824556
INFO:tensorflow:2017-12-08 23:15:50.600437: Step 3920: Validation accuracy = 14.0% (N=100)
INFO:tensorflow:2017-12-08 23:16:00.668603: Step 3930: Train accuracy = 72.0%
INFO:tensorflow:2017-12-08 23:16:00.670104: Step 3930: Cross entropy = 0.828033
INFO:tensorflow:2017-12-08 23:16:01.177465: Step 3930: Validation accuracy = 23.0% (N=100)
INFO

## Testing
1. Copy the two files named <b>output_...</b> from /tmp folder to the model folder.<br>
2. Run the following code.<br>
3. It will create a test_results.csv file containing the test results.

In [2]:
# Derivative Work: Stephen Houser & Mohammad Irfan [Bowdoin College, 2017]
# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import argparse
import sys
import os

import numpy as np
import tensorflow as tf

#This function loads the model that has been learned above
def load_graph(model_file):
    graph = tf.Graph()
    graph_def = tf.GraphDef()

    with open(model_file, "rb") as f:
        graph_def.ParseFromString(f.read())
    with graph.as_default():
        tf.import_graph_def(graph_def)

    return graph

#This function transforms a .jpg image file into its "tensor" form
def read_tensor_from_image_file(file_name, input_height=299, input_width=299,
				input_mean=0, input_std=255):
    input_name = "file_reader"
    output_name = "normalized"
    file_reader = tf.read_file(file_name, input_name)
    #Assuming .jpg file
    image_reader = tf.image.decode_jpeg(file_reader, channels = 3,
                                        name='jpeg_reader')
    float_caster = tf.cast(image_reader, tf.float32)
    dims_expander = tf.expand_dims(float_caster, 0);
    resized = tf.image.resize_bilinear(dims_expander, [input_height, input_width])
    normalized = tf.divide(tf.subtract(resized, [input_mean]), [input_std])
    sess = tf.Session()
    result = sess.run(normalized)

    return result

#This function reads the list of tags from the output_labels.txt file
def load_labels(label_file):
    label = []
    proto_as_ascii_lines = tf.gfile.GFile(label_file).readlines()
    for l in proto_as_ascii_lines:
        label.append(l.rstrip())
    return label

#This function tests each image from a test set of images
def test():
    model_file = "model/output_graph.pb"
    label_file = "model/output_labels.txt"
    input_name = "import/Mul"
    output_name = "import/final_result"

    input_height = 299
    input_width = 299
    input_mean = 0
    input_std = 255

    #Create an output file
    output_file = "test_results.csv"
    out_file_object = open(output_file, "wt")

    graph = load_graph(model_file)
    input_operation = graph.get_operation_by_name(input_name);
    output_operation = graph.get_operation_by_name(output_name);

    #Print column headers  
    labels = load_labels(label_file)
    #print("medal_file,", end='')
    out_file_object.write("file,")
    for l in labels:
        #print("{},".format(l), end='')
        out_file_object.write("{},".format(l))
    #print("ground_truth")
    out_file_object.write("ground_truth\n")

    for name in os.listdir("test_set"): #For each subfolder in test_set
        if name.find(".") >= 0: #Ignore hidden files like .DS_Store
            continue
        print("Folder name:", name)
        folder_name = os.path.join("test_set",name) #Create subfolder name
        for medal_file in os.listdir(folder_name): #For each file in that subfolder
            if medal_file.endswith(".jpg"):
                print("\t"+medal_file)
                t = read_tensor_from_image_file(os.path.join(folder_name, medal_file),
                                      input_height=input_height,
                                      input_width=input_width,
                                      input_mean=input_mean,
                                      input_std=input_std)

                with tf.Session(graph=graph) as sess:
                    results = sess.run(output_operation.outputs[0],
                              {input_operation.outputs[0]: t})
                results = np.squeeze(results) #results is an array of probabilities
                res = ",".join(map(str, results)) #map applies str to each probability in results
                #print("{},{},{}".format(medal_file, res, name))
                out_file_object.write("{},{},{}\n".format(medal_file, res, name))
    
    out_file_object.close()  

#Apply the model to the test_set
test() 

Folder name: animal
	1966.104.7.b.jpg
	1966.107.16.a.jpg
	1966.107.9.b.jpg
	1966.113.43.b.jpg
	1966.117.35.b.jpg
	1966.119.2.b.jpg
	1966.131.49.b.jpg
	1966.131.51.b.jpg
	1966.131.54.b.jpg
	1966.131.56.a.jpg
	1966.131.58.b.jpg
	1966.134.10.b.jpg
	1966.134.17.b.jpg
	1967.16.14.a.jpg
	1967.16.9.a.jpg
	1967.17.8.a.jpg
	1989.4.20.a.jpg
Folder name: blank
	1966.107.10.b.jpg
	1966.107.12.b.jpg
	1966.113.68.b.jpg
	1966.113.70.b.jpg
	1966.116.1.b.jpg
	1966.116.2.b.jpg
	1966.117.34.b.jpg
	1966.117.36.b.jpg
	1966.129.10.b.jpg
	1966.129.3.b.jpg
	1966.129.4.b.jpg
	1966.129.5.b.jpg
	1966.129.9.b.jpg
	1966.131.47.b.jpg
	1966.137.165.b.jpg
	1967.14.22.b.jpg
	1967.14.23.b.jpg
	1967.14.24.b.jpg
	1967.14.25.b.jpg
	1967.14.28.b.jpg
	1967.15.6.b.jpg
	1967.16.14.b.jpg
	1967.16.9.b.jpg
	1967.17.4.b.jpg
	1967.17.5.b.jpg
	1967.17.8.b.jpg
	1967.17.9.b.jpg
	1967.20.40.b.jpg
	1967.20.41.b.jpg
	1967.20.52.b.jpg
	1967.20.55.b.jpg
	1967.20.57.a.jpg
	1967.20.57.b.jpg
	1967.20.61.b.jpg
	1967.20.64.b.jpg
	1967.20.66.b.

	1966.137.28.b.jpg
	1967.20.66.a.jpg
	1989.4.11.b.jpg
	1989.4.12.b.jpg
	1989.4.13.b.jpg
	1989.4.19.b.jpg
Folder name: text
	1966.104.10.a.jpg
	1966.104.10.b.jpg
	1966.104.10.jpg
	1966.104.11.a.jpg
	1966.104.11.b.jpg
	1966.104.12.a.jpg
	1966.104.13.a.jpg
	1966.104.13.b.jpg
	1966.104.14.a.jpg
	1966.104.14.b.jpg
	1966.104.15.a.jpg
	1966.104.15.b.jpg
	1966.104.6.b.jpg
	1966.104.7.a.jpg
	1966.104.8.a.jpg
	1966.104.9.a.jpg
	1966.104.9.b.jpg
	1966.105.1.a.jpg
	1966.105.1.b.jpg
	1966.107.7.b.jpg
	1966.113.47.b.jpg
	1966.113.76.b.jpg
	1966.114.38.b.jpg
	1966.114.39.b.jpg
	1966.114.40.b.jpg
	1966.114.41.b.jpg
	1966.116.3.b.jpg
	1966.116.4.b.jpg
	1966.117.4.b.jpg
	1966.117.5.b.jpg
	1966.117.6.b.jpg
	1966.117.7.b.jpg
	1966.117.8.b.jpg
	1966.129.11.b.jpg
	1966.129.14.b.jpg
	1966.129.15.b.jpg
	1966.129.6.b.jpg
	1966.129.7.b.jpg
	1966.129.8.b.jpg
	1966.131.46.a.jpg
	1966.131.46.b.jpg
	1966.131.47.a.jpg
	1966.131.48.a.jpg
	1966.131.48.b.jpg
	1966.131.49.a.jpg
	1966.131.49.b.jpg
	1966.131.51.a.jpg
	196

# Questions 1

Write Python codes that reads in the test_results.csv file and computes (a) the overall test accuracy (in %) over the whole test set and (b) the test accuracy for each type of image (e.g., child).    



In [3]:
def test_accuracy(csv_file):
    #Read in file
    file_object = open(csv_file, "rt")
    all_text = file_object.read()
    file_object.close()
    lines = all_text.split("\n")
    #print(lines[1])
    
    #A. Compute overall test accuracy: 
        #Check whether category identified in ground_truth corresponds with highest probability 2 categories
            #Create dictionary assigning each probability value to its category
    
    accurate_count = 0
    headers = lines[0].split(",")
    categories = {}
    hit = {}
    miss ={}
    
    for i in range(1, (len(headers)-1)): #Assign hit and miss dictionary values to 0
        hit[headers[i]] = 0
        miss[headers[i]] = 0

    for line in lines[1:-1]: #Go line by line
        data = line.split(",")
        ground_truth = data[-1]
        #print(data[14])
        

        for i in range(1, (len(data)-1)): #Assign probability value to category header
            categories[headers[i]] = data[i]
        #print(categories)
        
        #Reverse sort dictionary keys
        sorted_categories = sorted(categories, key = categories.get, reverse = True)
        #print(sorted_categories)
        
        
        #Check whether highest value header matches ground_truth
        if sorted_categories[0] == ground_truth or sorted_categories[1] == ground_truth:
            
            #Log the number of accurate responses
            accurate_count += 1
            hit[ground_truth] += 1
        else:
            miss[ground_truth] += 1
    
        
    #Calculate overall percent accurate
   
    print("Overall test accuracy:", (accurate_count / (len(lines)-2) * 100), "%" )
    #print("Hit:", hit)
    #print("Miss:", miss)
    
    #B. Calculate accuracy for each category:
    print("Percent accuracy for each image type:")
    for header in hit:
        print(header,((hit[header] / (hit[header] + miss[header]))* 100), "%")
    
        
        
    

test_accuracy("test_results.csv")

Overall test accuracy: 54.757281553398066 %
Percent accuracy for each image type:
animal 58.82352941176471 %
blank 61.53846153846154 %
building 78.26086956521739 %
child 10.0 %
figure 22.58064516129032 %
group 36.76470588235294 %
mythical 20.689655172413794 %
object 6.666666666666667 %
plants 33.33333333333333 %
portrait 84.49612403100775 %
symbol 60.86956521739131 %
text 64.0 %


First test file results

Overall test accuracy: 54.17298937784521 % <br>
Percent accuracy for each image type:<br>
blank 71.15384615384616 %<br>
text 45.33333333333333 %<br>
symbol 65.21739130434783 %<br>
mythical 24.137931034482758 %<br>
animal 41.17647058823529 %<br>
child 20.0 %<br>
object 6.666666666666667 %<br>
figure 25.806451612903224 %<br>
group 41.17647058823529 %<br>
plants 33.33333333333333 %<br>
portrait_female 43.333333333333336 %<br>
building 78.26086956521739 %<br>
portrait 73.64341085271317 %<br>
portrait_male 70.17543859649122 %<br>

Second test file results 

Overall test accuracy: 54.757281553398066 %<br>
Percent accuracy for each image type:<br>
animal 58.82352941176471 %<br>
blank 61.53846153846154 %<br>
building 78.26086956521739 %<br>
child 10.0 %<br>
figure 22.58064516129032 %<br>
group 36.76470588235294 %<br>
mythical 20.689655172413794 %<br>
object 6.666666666666667 %<br>
plants 33.33333333333333 %<br>
portrait 84.49612403100775 %<br>
symbol 60.86956521739131 %<br>
text 64.0 %<br>

# Question 2
Do the following: 

    (i) Move the portrait_male and portrait_female subfolders from the training_set to your Desktop (so that they are no longer parts of the training set).
    (ii) Re-do the training and testing steps above.

Now repeat Question 1 on the new test_results.csv file. Comment on why you are you seeing a difference (if any) in test accuracy.

Increased accuracy: text, animal, portrait

Decreased accuracy: blank, symbol, mythical, child, figure, group

No change: object, plants, building

The overall percent accuracy only increased marginally.  While the accuracy for portraits increased from 73.6% to 84.5%, the accuracy for a number of other groups decreased.  It seems that very distinctive features like objects, plants, and buildings that were unlikely to either be confused with a portrait or also appear in a portrait were not affected.  It makes sense that the accuracy for portrait and text would increase because images from the portrait_woman and portrait_man folders were certain to also be tagged as portraits and were likely to also contain text.  Removing the competition for possible tags allowed a more accurate evaluation.  The other more ambiguous unrelated categories may have decreased in accuracy because there were fewer categories to distinguish images.

# Question 3 EC
Copy and paste the above code for testing below and modify it for automatically tagging the images in the test_set_2 folder, none of which has been tagged by anyone before.

    (i) Change the name of the output file to test_results_2.csv (so that you don't replace the previous test_results.csv file).
    
    (ii) Delete "ground_truth". We don't know the ground truth now as the images have not been tagged before.
    
    (ii) Change the folder name where the new test images are. This folder doesn't have any subfolder. Therefore, delete one for loop that goes over subfolders. Make necessary changes due to the deletion of that for loop.

In [ ]:
# Derivative Work: Stephen Houser & Mohammad Irfan [Bowdoin College, 2017]
# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import argparse
import sys
import os

import numpy as np
import tensorflow as tf

#This function loads the model that has been learned above
def load_graph(model_file):
    graph = tf.Graph()
    graph_def = tf.GraphDef()

    with open(model_file, "rb") as f:
        graph_def.ParseFromString(f.read())
    with graph.as_default():
        tf.import_graph_def(graph_def)

    return graph

#This function transforms a .jpg image file into its "tensor" form
def read_tensor_from_image_file(file_name, input_height=299, input_width=299,
				input_mean=0, input_std=255):
    input_name = "file_reader"
    output_name = "normalized"
    file_reader = tf.read_file(file_name, input_name)
    #Assuming .jpg file
    image_reader = tf.image.decode_jpeg(file_reader, channels = 3,
                                        name='jpeg_reader')
    float_caster = tf.cast(image_reader, tf.float32)
    dims_expander = tf.expand_dims(float_caster, 0);
    resized = tf.image.resize_bilinear(dims_expander, [input_height, input_width])
    normalized = tf.divide(tf.subtract(resized, [input_mean]), [input_std])
    sess = tf.Session()
    result = sess.run(normalized)

    return result

#This function reads the list of tags from the output_labels.txt file
def load_labels(label_file):
    label = []
    proto_as_ascii_lines = tf.gfile.GFile(label_file).readlines()
    for l in proto_as_ascii_lines:
        label.append(l.rstrip())
    return label

#This function tests each image from a test set of images
def test():
    model_file = "model/output_graph.pb"
    label_file = "model/output_labels.txt"
    input_name = "import/Mul"
    output_name = "import/final_result"

    input_height = 299
    input_width = 299
    input_mean = 0
    input_std = 255

    #Create an output file
    output_file = "test_results_2.csv"
    out_file_object = open(output_file, "wt")

    graph = load_graph(model_file)
    input_operation = graph.get_operation_by_name(input_name);
    output_operation = graph.get_operation_by_name(output_name);

    #Print column headers  
    labels = load_labels(label_file)
    #print("medal_file,", end='')
    out_file_object.write("file,")
    for l in labels:
        #print("{},".format(l), end='')
        out_file_object.write("{},".format(l))
    #print("ground_truth")
    #out_file_object.write("ground_truth\n")
        
    for name in os.listdir("test_set_2"): #For each subfolder in test_set
        if name.find(".") >= 0: #Ignore hidden files like .DS_Store
            continue
            
    for medal_file in os.listdir("test_set_2"): #For each file in test_set_2
        if medal_file.endswith(".jpg"):
            print("\t"+medal_file)
            t = read_tensor_from_image_file(os.path.join("test_set_2", medal_file),
                                  input_height=input_height,
                                  input_width=input_width,
                                  input_mean=input_mean,
                                  input_std=input_std)

            with tf.Session(graph=graph) as sess:
                results = sess.run(output_operation.outputs[0],
                          {input_operation.outputs[0]: t})
            results = np.squeeze(results) #results is an array of probabilities
            res = ",".join(map(str, results)) #map applies str to each probability in results
            #print("{},{},{}".format(medal_file, res, name))
            out_file_object.write("{},{},{}\n".format(medal_file, res, name))
    
    out_file_object.close()  

#Apply the model to the test_set
test() 

	1966.106.13.b.jpg
	1966.106.14.a.jpg
	1966.106.14.b.jpg
	1966.106.15.a.jpg
	1966.106.15.b.jpg
	1966.106.16.a.jpg
	1966.106.16.b.jpg
	1966.106.17.a.jpg
	1966.108.10.a.jpg
	1966.108.10.b.jpg
	1966.108.11.a.jpg
	1966.108.11.b.jpg
	1966.108.12.a.jpg
	1966.108.12.b.jpg
	1966.108.13.a.jpg
	1966.108.13.b.jpg
	1966.108.13.D.a.jpg
	1966.108.13.D.b.jpg
	1966.108.14.a.jpg
	1966.108.14.b.jpg
	1966.108.2.b.jpg
	1966.108.3.a.jpg
	1966.108.3.b.jpg
	1966.108.4.a.jpg
	1966.108.4.b.jpg
	1966.108.5.a.jpg
	1966.108.5.b.jpg
	1966.108.7.a.jpg
	1966.108.7.b.jpg
	1966.108.8.a.jpg
	1966.108.8.b.jpg
	1966.108.9.a.jpg
	1966.108.9.b.jpg
	1966.113.16.a.jpg
	1966.113.16.b.jpg
	1966.113.17.a.jpg
	1966.113.17.b.jpg
	1966.113.18.a.jpg
	1966.113.18.b.jpg
	1966.113.19.a.jpg
	1966.113.19.b.jpg
	1966.113.20.a.jpg
	1966.113.20.b.jpg
	1966.113.21.a.jpg
	1966.113.21.b.jpg
	1966.113.24.a.jpg
	1966.113.24.b.jpg
	1966.113.25.a.jpg
	1966.113.25.b.jpg
	1966.113.26.a.jpg
	1966.113.26.b.jpg
	1966.113.27.a.jpg
	1966.113.27.b.jpg
	1

# Question 4
What are your observations on the powers and limitations of machine learning?

Machine learning is an extremely powerful tool that is being used across a wide range of fields to produce new insights and automate tasks that we previously thought only humans could do.  Whereas it seemed that in the past, computers could only perform straightforward computation, machine learning gives them the ability to perform tasks that seem to require human insight and a genuine understanding of the world around us.  We must remember, though, that this is an illusion produced not by the increased humanity of machines but the increased sophistication of computational algorithms created by a certain subset of humans.  Who those humans are is of the utmost importance, as we have discussed, as they are the architects of a world coming to be increasingly driven by machine learning.  Human error and bias also comes into play in the training phase of machine learning, as the data set that we choose to input and the output that we choose to elicit has a substantial effect on the seemingly impartial, rational information that these algorithms give us.  Therefore must be careful to remember this: as machine learning algorithms take increasingly shape our world and the way in which we interact with and perceive it, they are only as accurate and objective as their creators. 

# Question 5
Anne Goodyear talked about how medals are everyday objects that represent power and have been used to influence publics. How does this relate to contemporary everyday objects that are symbols of power in the United States? Choose a specific one (could be a real or digital object) that interests you and discuss how it connects to the medals. (2 paragraphs)

Medals represent networks of power and influence: 1500s-1900

When I think of the most salient indicators of one's networks of power and influence today, two sets of words arise in my mind.  The first is money, cars, and girls; the second is likes and followers.  One of these seems decidedly anchored in the mid-20th century, the other in the 21st, though both retain their relevance today.  However, I would argue that the second is becoming far more powerful than the first.  In fact, this transition of power can be seen in the recent Harvey Weinstein scandal and its repercussions: while for decades men exerted their power over women as a self-reflexive assertion of their power and influence, online social media allowed women worldwide to gain power and influence through a different medium, and using the #metoo hashtag to attack and topple misogynistic power structures in the real world.  This demonstrates that a person's influence on social media is quickly becoming the locus of the greatest degree of social power.
    
Just as a person's follower count marks a person as influential (and therefore powerful) in our present society, the possession of medals with one's self-portrait marked a person as influential and powerful in past societies.  There are some major differences between these two mediums, however.  First of all, medals indicated power precisely because one had to be wealthy in order to have them created.  The medals were more of a symbol than a locus of power.  Social media followership, on the other hand, does not inherently require wealth, but instead the consistent output of content that the public finds valuable.  For that reason, social media followership is both a symbol of power and, even more importantly, the locus of power, as it is through this network that the person has the power to disseminate their thoughts and ideas.  This indicates a radical shift in our present society from a society driven by wealth to a society driven by social capital in the form of "influence," an asset that virtually anyone can forge, or lose, with the creation of a single piece of content.  The threshold to entry (and exit) is thus much lower than it has been throughout past centuries. 